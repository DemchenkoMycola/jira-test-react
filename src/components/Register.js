import React from 'react';
import { connect } from 'react-redux';
import { addUser } from '../actions/userActions';

class Register extends React.Component{

    constructor(props){
        super(props);
        this.state = {
            username: '',
            email: '',
            password: '',
        }

        this.handleName = this.handleName.bind(this);
        this.handleEmail = this.handleEmail.bind(this);
        this.handlePassword = this.handlePassword.bind(this);
        this.handleSubmit = this.handleSubmit.bind(this);
    }

    handleName(e){
        this.setState({
            username: e.target.value
        })
    }

    handleEmail(e){
        this.setState({
            email: e.target.value
        })
    }

    handlePassword(e){
        e.preventDefault();

        this.setState({
            password: e.target.value
        })
    }

    handleSubmit(e){
        e.preventDefault();
        this.props.addUser(this.state)
    }

    render(){
        return(
            <section className="register_form">
                <legend>Register</legend>
                <label>Username</label>
                <input 
                    onChange={this.handleName} 
                    type="text" name="username"
                    placeholder="username" />
                <label>Email</label>
                <input 
                    onChange={this.handleEmail}
                    type="email" 
                    name="email" 
                    placeholder="email" />
                <label>Password</label>
                <input 
                    onChange={this.handlePassword}
                    type="password" 
                    name="passowrd" 
                    placeholder="password"/>
                <button onClick={this.handleSubmit}>register</button>
            </section>
        )
    }
}

const mapDispatchToProps = (dispatch) => {
    return {
        addUser: user => dispatch(addUser(user))
    }
}

export default connect(null, mapDispatchToProps)(Register);