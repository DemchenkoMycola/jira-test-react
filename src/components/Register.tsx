import React from 'react';
import { connect } from 'react-redux';
import { addUser } from '../actions/userActions';

interface User {
    name: string,
    email: string,
    password: string
}

interface Props {
    users: object[],
    addUser: (user: any) => void,
    handleLogin: () => void
}

interface State {
    errors: string,
    user: User,
    messages: string
}

type InputEvent = React.ChangeEvent<HTMLInputElement>;
type ButtonEvent = React.MouseEvent<HTMLButtonElement>;

class Register extends React.Component<Props, State>{

    constructor(props: any){
        super(props);
        this.state = {
            user: {
                name: '',
                email: '',
                password: '',
            },
            errors: '',
            messages: ''
        }

        this.handleName = this.handleName.bind(this);
        this.handleEmail = this.handleEmail.bind(this);
        this.handlePassword = this.handlePassword.bind(this);
        this.handleSubmit = this.handleSubmit.bind(this);
    }

    handleName(e: InputEvent){        
        let username = e.target.value
        if(this.props.users.find((user: any) => user.name === username)){
            this.setState({
                errors: `Username exist ${this.state.errors}`,
                user: {...this.state.user, name: username},
            })
        }else{
            this.setState({
                user: {...this.state.user, name: username},
                errors: ''
            })
        }
    }

    handleEmail(e: InputEvent){

        let email = e.target.value;

        if(this.props.users.find((user: any) => user.email === email)){
            this.setState({
                errors: `Chose another email ${this.state.errors}`,
                user: {...this.state.user, email: e.target.value}
            })
        }else{
            this.setState({
                user: {...this.state.user, email: e.target.value},
                errors: ''
            })
        }

    }

    handlePassword(e: InputEvent){
        e.preventDefault();
        this.setState({
            user: {...this.state.user, password: e.target.value}
        })
    }

    handleSubmit(e: ButtonEvent){
        if(this.state.errors !== ''){
            this.setState({
                errors: `Please fix errors  \n ${this.state.errors}` 
            })
        }else{
            e.preventDefault();
            this.props.addUser(this.state.user);
            this.setState({
                errors: '',
                messages: 'Thanks for register',
                user: {
                    name: '',
                    email: '',
                    password: '',
                }
            });
            this.props.handleLogin();
        }
    }

    render(){
        return(
            <section className="register_form">
                <legend>Register</legend>
                <span className="errors">{this.state.errors}</span>
                <span className="messages">{this.state.messages}</span>
                <label>Username</label>
                <input 
                    className="input"
                    onChange={this.handleName} 
                    type="text" 
                    name="username"
                    value={this.state.user.name}
                    placeholder="username" />
                <label>Email</label>
                <input 
                    className="input"                    
                    onChange={this.handleEmail}
                    type="email" 
                    name="email"
                    value={this.state.user.email}
                    placeholder="email" />
                <label>Password</label>
                <input 
                    className="input"            
                    onChange={this.handlePassword}
                    type="password" 
                    name="passowrd" 
                    value={this.state.user.password}
                    placeholder="password"/>
                <button 
                className="button register_button"
                disabled={this.state.user.name === ''}
                onClick={this.handleSubmit}>register</button>
            </section>
        )
    }
}

const mapDispatchToProps = (dispatch: any) => {
    return {
        addUser: (user: User) => dispatch(addUser(user))
    }
}

const mapStateToProps = (state: any) => {
    return{
        users: state.users
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(Register);