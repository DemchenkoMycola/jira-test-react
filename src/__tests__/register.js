import React from 'react';
import Register from '../components/Register'
import renderer from 'react-test-renderer';
import { createStore } from 'redux';

describe('register', () => {

    const store = createStore(() => ({}));

    test('render register form', () => {
        const login = renderer.create(<Register store={store} />);
        expect(login).toMatchSnapshot();
    });

});