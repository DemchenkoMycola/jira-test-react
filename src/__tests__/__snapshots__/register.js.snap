// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`register render register form 1`] = `
<section
  className="register_form"
>
  <legend>
    Register
  </legend>
  <span
    className="errors"
  >
    
  </span>
  <span
    className="messages"
  >
    
  </span>
  <label>
    Username
  </label>
  <input
    className="input"
    name="username"
    onChange={[Function]}
    placeholder="username"
    type="text"
    value=""
  />
  <label>
    Email
  </label>
  <input
    className="input"
    name="email"
    onChange={[Function]}
    placeholder="email"
    type="email"
    value=""
  />
  <label>
    Password
  </label>
  <input
    className="input"
    name="passowrd"
    onChange={[Function]}
    placeholder="password"
    type="password"
    value=""
  />
  <button
    className="button register_button"
    disabled={true}
    onClick={[Function]}
  >
    register
  </button>
</section>
`;
